lucille-core/src/main/java/com/kmwllc/lucille/util/FileContentFetcher.java

```diff
@@ -0,0 +1,141 @@
+package com.kmwllc.lucille.util;
+
+import com.kmwllc.lucille.connector.storageclient.StorageClient;
+import com.kmwllc.lucille.core.ConnectorException;
+import com.kmwllc.lucille.core.StageException;
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.URI;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Responsible for working with StorageClients to repeatedly fetch the contents of String paths, determining the appropriate
+ * storage provider.
+ */
+public class FileContentFetcher {
+  private static final Logger log = LoggerFactory.getLogger(FileContentFetcher.class);
+
+  private final Map<String, StorageClient> availableClients;
+
+  /**
+   * Creates a FileContentFetcher that will only fetch files from the classpath or the local file system. Be sure to call startup
+   * and shutdown as appropriate.
+   */
+  public FileContentFetcher() {
+    this.availableClients = StorageClient.clientsFromCloudOptions(Map.of());
+  }
+
+  /**
+   * Creates a FileContentFetcher that will fetch files from the classpath, the local file system, and any cloud file systems
+   * (Azure, S3, Google) which have the necessary options provided in the given options map. Be sure to call startup and shutdown
+   * as appropriate.
+   */
+  public FileContentFetcher(Map<String, Object> cloudOptions) {
+    this.availableClients = StorageClient.clientsFromCloudOptions(cloudOptions);
+  }
+
+  /**
+   * Initialize the storage clients associated with this FileContentFetcher. Calls shutdown on all clients and throws an Exception
+   * if an error occurs while initializing any of the StorageClients.
+   */
+  public void startup() throws StageException {
+    for (StorageClient client : availableClients.values()) {
+      try {
+        client.init();
+      } catch (ConnectorException e) {
+        shutdown();
+
+        throw new StageException("Unable to initialize StorageClient.", e);
```

Comment:
not sure this should be a StageException as this method could be called outside a stage...